### ISSUE TITLE
Storing Dio API response fails

### ISSUE THREAD

**Issue by Prn-Ice** [fstof/flutter_json_store#1]
I've got an app that uses the [dio](https://pub.dev/packages/dio) plugin for get and push requests.
I would like to use your plugin to store caches of the response from those calls,  but when I attempt to store it I get `Converting object to an encodable object failed: Instance of 'Response<dynamic>'`
 
This is how I encode the data
```dart
class ResponseModel {
  ResponseModel({@required this.value});
  ResponseModel.fromJson(Map<String, dynamic> json) : value = json['value'];

  Response<dynamic> value;

  Map<String, dynamic> toJson() => <String, dynamic>{'value': value};
}

JsonStore jsonStore = JsonStore(dbName: 'popular_places');
  Response<dynamic> initialData;
  ResponseModel responseModel;

  Future<void> initialiseCache() async {
    final Response<dynamic> response =
        await NetworkHelper(kAPIPopularPlaces).getData();
    responseModel = ResponseModel(value: response);
    await jsonStore.setItem('places', responseModel.toJson());
    setState(() {});
  }

```

The full error:
```
Launching lib/main.dart on Android SDK built for x86 in debug mode...
Initializing gradle...
Resolving dependencies...
Running Gradle task 'assembleDebug'...
Built build/app/outputs/apk/debug/app-debug.apk.
Installing build/app/outputs/apk/app.apk...
I/Choreographer( 4663): Skipped 34 frames!  The application may be doing too much work on its main thread.
D/EGL_emulation( 4663): eglMakeCurrent: 0xe9705720: ver 3 1 (tinfo 0xd12f3e90)
Syncing files to device Android SDK built for x86...
D/EGL_emulation( 4663): eglMakeCurrent: 0xe7d6a4a0: ver 3 1 (tinfo 0xe6adbdd0)
I/flutter ( 4663): showOnBoarding is true
D/skia    ( 4663): Shader compilation error
D/skia    ( 4663): ------------------------
D/skia    ( 4663): Errors:
D/skia    ( 4663): 
D/skia    ( 4663): Shader compilation error
D/skia    ( 4663): ------------------------
D/skia    ( 4663): Errors:
D/skia    ( 4663): 
D/skia    ( 4663): Shader compilation error
D/skia    ( 4663): ------------------------
D/skia    ( 4663): Errors:
D/skia    ( 4663): 
D/skia    ( 4663): Shader compilation error
D/skia    ( 4663): ------------------------
D/skia    ( 4663): Errors:
D/skia    ( 4663): 
D/skia    ( 4663): Shader compilation error
D/skia    ( 4663): ------------------------
D/skia    ( 4663): Errors:
D/skia    ( 4663): 
D/skia    ( 4663): Shader compilation error
D/skia    ( 4663): ------------------------
D/skia    ( 4663): Errors:
D/skia    ( 4663): 
E/flutter ( 4663): [ERROR:flutter/lib/ui/ui_dart_state.cc(148)] Unhandled Exception: Instance of 'StorageException'
E/flutter ( 4663): #0      JsonStore.setItem (package:json_store/src/json_store.dart:121:7)
E/flutter ( 4663): <asynchronous suspension>
E/flutter ( 4663): #1      _PopularPlacesState.initialiseCache (package:riv_app/screens/main/popular_places.dart:30:21)
E/flutter ( 4663): <asynchronous suspension>
E/flutter ( 4663): #2      _PopularPlacesState.initState (package:riv_app/screens/main/popular_places.dart:46:5)
E/flutter ( 4663): #3      StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:4068:58)
E/flutter ( 4663): #4      ComponentElement.mount (package:flutter/src/widgets/framework.dart:3919:5)
E/flutter ( 4663): #5      Element.inflateWidget (package:flutter/src/widgets/framework.dart:3101:14)
E/flutter ( 4663): #6      Element.updateChild (package:flutter/src/widgets/framework.dart:2904:12)
E/flutter ( 4663): #7      SingleChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:5127:14)
E/flutter ( 4663): #8      Element.inflateWidget (package:flutter/src/widgets/framework.dart:3101:14)
E/flutter ( 4663): #9      Element.updateChild (package:flutter/src/widgets/framework.dart:2904:12)
E/flutter ( 4663): #10     SingleChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:5127:14)
E/flutter ( 4663): #11     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3101:14)
E/flutter ( 4663): #12     Element.updateChild (package:flutter/src/widgets/framework.dart:2904:12)
E/flutter ( 4663): #13     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3961:16)
E/flutter ( 4663): #14     Element.rebuild (package:flutter/src/widgets/framework.dart:3738:5)
E/flutter ( 4663): #15     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:3924:5)
E/flutter ( 4663): #16     ComponentElement.mount (package:flutter/src/widgets/framework.dart:3919:5)
E/flutter ( 4663): #17     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3101:14)
E/flutter ( 4663): #18     Element.updateChild (package:flutter/src/widgets/framework.dart:2904:12)
E/flutter ( 4663): #19     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3961:16)
E/flutter ( 4663): #20     Element.rebuild (package:flutter/src/widgets/framework.dart:3738:5)
E/flutter ( 4663): #21     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:3924:5)
E/flutter ( 4663): #22     ComponentElement.mount (package:flutter/src/widgets/framework.dart:3919:5)
E/flutter ( 4663): #23     ParentDataElement.mount (package:flutter/src/widgets/framework.dart:4314:11)
E/flutter ( 4663): #24     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3101:14)
E/flutter ( 4663): #25     Element.updateChild (package:flutter/src/widgets/framework.dart:2904:12)
E/flutter ( 4663): #26     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3961:16)
E/flutter ( 4663): #27     Element.rebuild (package:flutter/src/widgets/framework.dart:3738:5)
E/flutter ( 4663): #28     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:3924:5)
E/flutter ( 4663): #29     StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:4088:11)
E/flutter ( 4663): #30     ComponentElement.mount (package:flutter/src/widgets/framework.dart:3919:5)
E/flutter ( 4663): #31     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3101:14)
E/flutter ( 4663): #32     Element.updateChild (package:flutter/src/widgets/framework.dart:2904:12)
E/flutter ( 4663): #33     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3961:16)
E/flutter ( 4663): #34     Element.rebuild (package:flutter/src/widgets/framework.dart:3738:5)
E/flutter ( 4663): #35     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:3924:5)
E/flutter ( 4663): #36     ComponentElement.mount (package:flutter/src/widgets/framework.dart:3919:5)
E/flutter ( 4663): #37     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3101:14)
E/flutter ( 4663): #38     Element.updateChild (package:flutter/src/widgets/framework.dart:2904:12)
E/flutter ( 4663): #39     SliverMultiBoxAdaptorElement.updateChild (package:flutter/src/widgets/sliver.dart:1181:36)
E/flutter ( 4663): #40     SliverMultiBoxAdaptorElement.createChild.<anonymous closure> (package:flutter/src/widgets/sliver.dart:1166:20)
E/flutter ( 4663): #41     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2328:19)
E/flutter ( 4663): #42     SliverMultiBoxAdaptorElement.createChild (package:flutter/src/widgets/sliver.dart:1159:11)
E/flutter ( 4663): #43     RenderSliverMultiBoxAdaptor._createOrObtainChild.<anonymous closure> (package:flutter/src/rendering/sliver_multi_box_adaptor.dart:354:23)
E/flutter ( 4663): #44     RenderObject.invokeLayoutCallback

```

**Comment by fstof** [comment]
@Prn-Ice 
So I don't have any hands on experince with `dio`. But I did have a look at their documentation.

First of all. You are trying to store the complete dio `Response` object in the storage. This is a problem as the dio `Response` object is not parsable to a `Map` (which is what json_store requires)

### TLDR
So the default `ResponseType` of dio is `json` I'm gonna go ahead and assume this is the one you are using.
If I look at the examle [here](https://github.com/flutterchina/dio/blob/master/example/options.dart) you will see that the dio's `ResponseType.json` will parse the response body from the resource to a `Map` (this is good news) so we should be able to just take that map and store it into json_store.

### Try this
In your case you could actually get rid of your `ResponseModel` class and just use the response body directly.
```dart
    final Response<Map> response = await NetworkHelper(kAPIPopularPlaces).getData();
    await jsonStore.setItem('places', response.data);
```
I tried the following on my side and worked just fine:
```dart
    Response<Map> response = await dio.get('https://swapi.co/api/people/1');
    await jsonStore.setItem('luke', response.data);
    var luke = await jsonStore.getItem('luke');
    print(luke);
```

If you really need the `ResponseModel` class you should maybe change the type of `value` from `Response<dynamic>` to just `Map` which will hold the value from `Response.data`

Good luck and shout if there's still issues

**Comment by Prn-Ice** [comment]
Thanks, that would be the way to go but the API returns a list of maps.

**Comment by fstof** [comment]
If you have a list of maps, you could store it in one of two ways.
1. Iterate over the list and add them one by one. This is useful if you ever want to access a single element from the list later
2. Wrap your list inside a map `jsonStore.setItem('places', {'items': response.data});`

**Comment by Prn-Ice** [comment]
That works thanks.
